### PR TITLE
Exercise 5

### PR DESCRIPTION
**Pull request by NicolasTsc**
Unser Review : https://github.com/Systems-Development-and-Frameworks/AlElMo/pull/8#pullrequestreview-564692795
Leider war die Gruppe zum Zeitpunkt des Reviews noch nicht zu 100% fertig, allerdings wollten wir das Review schonmal fertig haben.

### CODE REVIEW

**Review by roschaefer — CHANGES_REQUESTED** [review]
so far so good

This is an intermediary feedback, so no grading here. I think it looks good already. The biggest problem that I can see right now is that you run completely different code in your tests vs. during production. See this forum post in our moodle where I tried to explain the different options: https://moodle.htw-berlin.de/mod/forum/discuss.php?d=134501#p198015

You can run tests against Neo4J by cleaning the database for each test case or you could mock the neo4j driver response.

Also, play a little with Cypher. You can DRY up your code a lot. Don't be afraid to find and mutate sub-graphs in your database in one statement.

File: backend/README.md [hunk 1]
```diff
@@ -1,6 +1,27 @@
 # Apollo Backend
 
-## Mandatory for running tests ´
+
+## Why we decided to use the Neo4J architecture
```
Comment (moengers) [inline backend/README.md:4]: ⭐ For explaining WHY you have chosen the scenario in the README.md.

File: backend/README.md [hunk 2]
```diff
@@ -1,6 +1,27 @@
...
+
+We are both interested in learning new things in this course. That`s why it was nice to learn a new way to transfer data via GraphQL instead of
+the commmonly used REST-APIs.
+To learn even more new frameworks, tools, etc. we decided to choose the Neo4J architecture, because we were already interested in learning graph databases.
+Furthermore the presentation from William in the lecture was very informative and increased our interes in graph databases further.
+These are the reasons why we use the Neo4J architecture for this exercise.
+
+## Installation guide for Apollo Backend + Neo4J
```
Comment (moengers) [inline backend/README.md:12]: ⭐ For choosing a scenario and writing installation instructions in README.md.

File: backend/.env.template [hunk 3]
```diff
@@ -1 +1,6 @@
-JWTSECRET=yourjwtsecret
\ No newline at end of file
+JWTSECRET=yourjwtsecret
```
Comment (moengers) [inline backend/.env.template:1]: ⭐ For not committing sensitive secrets (e.g. API keys) unencrypted to the repository.

File: backend/src/index.spec.ts [hunk 4]
```diff
@@ -1,176 +1,183 @@
...
+const resolvers = Resolvers({ subschema: schema })
+const stichedSchema = stitchSchemas({
+  subschemas: [schema],
+  typeDefs,
+  resolvers
+})
+const dataSource = new Neo4JDataSource()
+let driver
+let testClient : ApolloServerTestClient
+const setupServerAndReturnTestClient = async (testToken?) => {
```
Comment (moengers) [inline backend/src/index.spec.ts:37]: ⭐ ⭐ Your software tests are free of side effects.

File: backend/src/resolvers.ts [hunk 5]
```diff
@@ -1,25 +1,70 @@
...
+        schema: subschema,
+        operation: 'query',
+        fieldName: 'User',
+        context,
+        info
+      })
+      return user
     }
   },
   Mutation: {
```
Comment (moengers) [inline backend/src/resolvers.ts:26]: ⭐ ⭐ Data created in mutations are persisted and your queries and mutations still work as expected.

**Review by moengers — APPROVED** [review]
Sieht sehr gut aus!
Wir haben nichts zu beanstanden.

Zu dem Neo4J Teil können wir euch leider nicht viel sagen, da wir uns mit GraphCMS beschäftigt haben.

**Review by moengers — APPROVED** [review]
Ausversehen auf den Button gekommen, dass eine neue review angefragt wird.
Hier ist nichts weiter sonst.

**Review by medizinmensch — APPROVED** [review]
Hi Felni!

For me it looks like you fixed @roschaefer's biggest concern (running different code in tests vs. production).
I especially liked your README as well as your very clear and nicely readable test cases! 👏 

Thanks also to @moengers for his review!

To conclude:

:star: For choosing a scenario   and writing installation instructions in README.md.
:star:   For explaining WHY you have chosen the scenario in the README.md.
:star: For not committing   sensitive secrets (e.g. API keys) unencrypted to the repository.
:star:   :star: Data created in mutations are persisted and your queries and mutations   still work as expected.
:star: :star: Your software   tests are free of side effects.
:star:   Your database is never left in an invalid state.
:star: You deny the client to   call mutations and queries of your subschema directly.
:star:   For requesting a review and reviewing another team's PR.

That makes a total of 10/10 :star:

Todo

**Pull request merged**